### PR TITLE
feat(docker): add depType for multi-stage Dockerfiles

### PR DESCRIPTION
**Pull request by Dmitry9**
[Issue](https://github.com/renovatebot/renovate/issues/2364)
Because of Dockerfile multiline string will have last `FROM` always after all other `FROM` occurrences in this [loop](https://github.com/Dmitry9/renovate/blob/50ac46687e104e856f6613a75fb08e3558c8492c/lib/manager/dockerfile/extract.js#L66) we can assign every dependancy depType `stage` and when finalizing just reassign the last element of [deps](https://github.com/Dmitry9/renovate/blob/50ac46687e104e856f6613a75fb08e3558c8492c/lib/manager/dockerfile/extract.js#L126) array. The iterations of the `for loop` and `deps` array getting new elements are synchronous actions (e.g. the last element of `deps` will correspond to the bottom `FROM` of the Dockerfile content passed as a string to the `extractPackageFile` function). 
While testing on the real repo it has unexpected results:
test1
test2
I expected golang to be skipped (my guess it is not yet implemented, or because of apine:latest -> skip reason ->unsupported-value`).
this is an extract from debug.log:
```json
"config": {
    "dockerfile": [
        {
            "packageFile": "Dockerfile",
            "manager": "dockerfile",
            "deps": [
                {
                    "depName": "golang",
                    "currentFrom": "golang:1.7.3",
                    "currentDepTag": "golang:1.7.3",
                    "currentValue": "1.7.3",
                    "datasource": "docker",
                    "depType": "stage", <------ arrow not from debug.log
                    "lineNumber": 1,
                    "fromPrefix": "FROM",
                    "fromSuffix": "as builder",
                    "updates": [
                        {
                            "fromVersion": "1.7.3",
                            "toVersion": "1.12.0",
                            "newValue": "1.12.0",
                            "newMajor": 1,
                            "newMinor": 12,
                            "updateType": "minor",
                            "isSingleVersion": true,
                            "newVersion": "1.12.0"
                        }
                    ],
                    "warnings": [],
                    "sourceUrl": null,
                    "dockerRegistry": "https://index.docker.io",
                    "dockerRepository": "library/golang",
                    "releases": [
                        "1.7.3",
                        "..............",
                        "1.12.0"
                    ]
                },
                {
                    "depName": "alpine",
                    "currentFrom": "alpine:latest",
                    "currentDepTag": "alpine:latest",
                    "currentValue": "latest",
                    "datasource": "docker",
                    "depType": "final", <------ arrow not from debug.log 
                    "lineNumber": 7,
                    "fromPrefix": "FROM",
                    "fromSuffix": "",
                    "updates": [],
                    "warnings": [],
                    "skipReason": "unsupported-value"
                }
            ]
        }
    ],
}
```
Closes # 2364

### CODE REVIEW

**Comment by Dmitry9** [comment]
@rarkins please, take a look into a description and see large comment about enums. After your feedback, if I am on the correct way I will add tests.

File: lib/manager/dockerfile/extract.js [hunk 1]
```diff
@@ -1,3 +1,23 @@
...
+ * I think it is good Idea to move up file hierarchy all primitives
+ * like strings literals, booleans and numbers.
+ * Store enums in one place (like Android framework demands)
+ * will enable you to make changes easily in the future.
+ * It will be nice to declare [multiStageDockerBuildEnums] (and in the
+ * long term all other strings literals) inside the
+ * enums responsible file of the project and then import
+ * here as [m], then access value in concise way e.g. [m.STAGE].
+ * Just by inspecting enums file developer will be able to quick
+ * understand all possible states of entity without scouring the project.
```
Comment (rarkins) [inline lib/manager/dockerfile/extract.js]: On the other hand, if constants/enums are used in only one place in the repo like here, it's nice to have it in one place so a developer doesn't need to go off and find it in a separate file. I usually prefer to centralise definitions only when they are used from multiple places.

File: test/manager/ansible/__snapshots__/extract.spec.js.snap [hunk 2]
```diff
@@ -9,6 +9,7 @@ Array [
     "currentValue": undefined,
     "datasource": "docker",
     "depName": "busybox",
+    "depType": "stage",
```
Comment (rarkins) [inline test/manager/ansible/__snapshots__/extract.spec.js.snap]: `depType` shouldn't apply to ansible - only to `dockerfile`

File: test/manager/docker-compose/__snapshots__/extract.spec.js.snap [hunk 3]
```diff
@@ -9,6 +9,7 @@ Array [
     "currentValue": "alpine",
     "datasource": "docker",
     "depName": "quay.io/something/redis",
+    "depType": "stage",
```
Comment (rarkins) [inline test/manager/docker-compose/__snapshots__/extract.spec.js.snap]: depType not applicable here either

**Review by rarkins — CHANGES_REQUESTED** [review]
depType should be added only for Dockerfile and not for ansible, docker-composer, kubernetes, etc.

**Comment by Dmitry9** [comment]
@rarkins please say, should these [lines](https://github.com/renovatebot/renovate/blob/401b03850c517d9e72a31e0ea98b7c66fd3a3c83/lib/manager/dockerfile/extract.js#L55) be removed from source and tests now?
skip1
Also theese lines:
skip2

File: test/manager/ansible/__snapshots__/extract.spec.js.snap (hunk 2 above)
Comment (Dmitry9) [reply test/manager/ansible/__snapshots__/extract.spec.js.snap]: done

File: test/manager/docker-compose/__snapshots__/extract.spec.js.snap (hunk 3 above)
Comment (Dmitry9) [reply test/manager/docker-compose/__snapshots__/extract.spec.js.snap]: done

**Comment by rarkins** [comment]
@Dmitry9 no, I think the current logic remains OK. We still will renovate stages, now we just give control to users who may wish to apply one rule for stage and another for final

**Comment by Dmitry9** [comment]
@rarkins I made changes requested and added a test.

**Pull request merged**